### ISSUE TITLE
Postgraphile docker container hangs after 3-4 hours of working

### ISSUE THREAD

**Issue by speller** [graphile/postgraphile#1252]
I'm submitting a ... 

* [x] bug report

PostGraphile version: 4.6.0

Minimal SQL file that can be loaded into a clean database: any database

Steps to reproduce:

1. Create the following files:

`Dockerfile`
```Dockerfile
FROM graphile/postgraphile:latest
ADD .postgraphilerc.js /postgraphile/ 
```
`.postgraphilerc.js`
```JavaScript
module.exports.options = {
  host: '0.0.0.0',
  connection: process.env.PG_CONNECTION_STRING,
  ownerConnection: process.env.PG_OWNER_CONNECTION_STRING,
  schema: [process.env.PG_SCHEMA],
  defaultRole: process.env.PG_DEFAULT_ROLE,
  dynamicJson: true,
  retryOnInitFail: true,
  disableGraphiql: false,
  enhanceGraphiql: true,
  live: true,
  watch: true,
}
```

2. Build and run this container:

```bash
> docker build -t test/gql:latest .
> export DEBUG="postgraphile:postgres"
> docker run -d -it \
  -p 5001:5000 \
  -e PG_CONNECTION_STRING=postgres://helpdesk-local-root:123456@host.docker.internal:5433/helpdesk-local \
  -e PG_OWNER_CONNECTION_STRING=postgres://helpdesk-super:123456@host.docker.internal:5433/helpdesk-local \
  -e PG_SCHEMA=helpdesk_public \
  -e PG_DEFAULT_ROLE=sys_anonymous \
  -e DEBUG="$DEBUG" \
  --name test-graphql \
  test/gql:latest 
```
3. Wait 3-4 hours (doesn't depend on presence of queries)

Current behavior:

- Errors in container console:
> Error occurred for PG watching client; reconnecting in 2 seconds. Connection terminated unexpectedly
>  postgraphile:postgres listen postgraphile_watch +6m
>Error occurred for PG watching client; reconnecting in 2 seconds. Connection terminated unexpectedly
- Postgraphile hangs and doesn't respond to incoming queries at all (timeout). Only container restart helps.
- At the same time, the `/` url responds and shows the default express error: `Cannot GET /`
- The following command from inside the container works perfectly:
```bash
/postgraphile # nc -zv host.docker.internal 5433
host.docker.internal (192.168.65.2:5433) open
```

Expected behavior:

Postgraphile works normally.

**Comment by benjie** [comment]
> Wait 3-4 hours (doesn't depend on presence of queries)

So this 3-4 hour timeout happens whether or not there were queries say 30 minutes before the error occurs? Or it relies on there having been no queries?

---

I don't think it matters, but what is your host OS, PostgreSQL version, and did your computer go to sleep during this time?

**Comment by speller** [comment]
I actually run two containers - one is test (described in the issue) and second is working one, very similar (with JWT and few tunes). They all were pointing to the same db. The test and the working one hanged at the same time (and were started almost at the same time). The working one was actively used by a client that I'm developing. Suddenly queries from this client stopped responding (last successful query was executed ~5-10 mins before). I checked both containers and they both hanged.

My computer didn't go to sleep during the testing time.

**Comment by benjie** [comment]
Have written this script to attempt reproduction:

```bash
#!/usr/bin/env bash

set -e

docker stop speller-graphql || true
docker rm speller-graphql || true


cat > Dockerfile <<HERE
FROM graphile/postgraphile:latest
ADD .postgraphilerc.js /postgraphile/ 
HERE

cat > .postgraphilerc.js <<HERE
module.exports.options = {
  host: '0.0.0.0',
  connection: process.env.PG_CONNECTION_STRING,
  ownerConnection: process.env.PG_OWNER_CONNECTION_STRING,
  schema: [process.env.PG_SCHEMA],
  defaultRole: process.env.PG_DEFAULT_ROLE,
  dynamicJson: true,
  retryOnInitFail: true,
  disableGraphiql: false,
  enhanceGraphiql: true,
  live: true,
  watch: true,
}
HERE

psql -X -v ON_ERROR_STOP=1 postgres <<HERE
drop database if exists "helpdesk-local";
drop role if exists "helpdesk-local-root";
drop role if exists "helpdesk-super";
drop role if exists "sys_anonymous";
create role "sys_anonymous";
create role "helpdesk-super" with login password '123456' superuser;
create role "helpdesk-local-root" with login password '123456';
grant "sys_anonymous" to "helpdesk-local-root";
create database "helpdesk-local" owner "helpdesk-local-root";
HERE

psql -X1 -v ON_ERROR_STOP=1 helpdesk-local <<HERE
create schema helpdesk_public;
grant usage on schema helpdesk_public to "sys_anonymous";
create table helpdesk_public.foo (
  id serial primary key,
  text text
);
grant select on helpdesk_public.foo to "sys_anonymous";
insert into helpdesk_public.foo (text) values ('A'), ('B'), ('C');
HERE

docker pull graphile/postgraphile

docker build -t speller/gql:latest .

export DEBUG="postgraphile:postgres"

docker run -d -it \
  -p 5001:5000 \
  -e PG_CONNECTION_STRING=postgres://helpdesk-local-root:123456@172.17.0.1:5432/helpdesk-local \
  -e PG_OWNER_CONNECTION_STRING=postgres://helpdesk-super:123456@172.17.0.1:5432/helpdesk-local \
  -e PG_SCHEMA=helpdesk_public \
  -e PG_DEFAULT_ROLE=sys_anonymous \
  -e DEBUG="$DEBUG" \
  --name speller-graphql \
  speller/gql:latest

echo "Waiting 10 seconds for Docker to come up"
sleep 10

for I in `seq 0 20 300`; do
  echo
  echo
  echo
  echo
  date
  echo "Slept ${I} minutes, running request"
  echo
  echo

  curl -i -X POST -H 'Content-Type: application/graphql' -d '{fooById(id: 1){id text}}' http://localhost:5001/graphql

  # Wait 20 minutes
  sleep 1200
done;

docker stop speller-graphql
docker rm speller-graphql

psql -X -v ON_ERROR_STOP=1 postgres <<HERE
drop database if exists "helpdesk-local";
drop role if exists "helpdesk-local-root";
drop role if exists "helpdesk-super";
drop role if exists "sys_anonymous";
HERE
```

We'll see what it says.

```
./script.sh 
speller-graphql
speller-graphql
DROP DATABASE
DROP ROLE
DROP ROLE
DROP ROLE
CREATE ROLE
CREATE ROLE
CREATE ROLE
GRANT ROLE
CREATE DATABASE
CREATE SCHEMA
GRANT
CREATE TABLE
GRANT
INSERT 0 3
Using default tag: latest
latest: Pulling from graphile/postgraphile
Digest: sha256:ae708791606842da2aaa8587179a0505c2be1c09820386de471df5319ba7bcc7
Status: Image is up to date for graphile/postgraphile:latest
docker.io/graphile/postgraphile:latest
Sending build context to Docker daemon  6.144kB
Step 1/2 : FROM graphile/postgraphile:latest
 ---> 8c2f0665a259
Step 2/2 : ADD .postgraphilerc.js /postgraphile/
 ---> Using cache
 ---> f9b87b3c8cc0
Successfully built f9b87b3c8cc0
Successfully tagged speller/gql:latest
d32c71dc2bd55d384c0332bf45ef4d784f2c1ace7f4f7abec655855286a02902
Waiting 10 seconds for Docker to come up




Mon  6 Apr 13:06:32 BST 2020
Slept 0 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 12:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}
```

I do hope your reproduction steps were sufficient; you didn't spec a particular database shape so I've just created one table.

**Comment by benjie** [comment]
```
./script.sh 
speller-graphql
speller-graphql
DROP DATABASE
DROP ROLE
DROP ROLE
DROP ROLE
CREATE ROLE
CREATE ROLE
CREATE ROLE
GRANT ROLE
CREATE DATABASE
CREATE SCHEMA
GRANT
CREATE TABLE
GRANT
INSERT 0 3
Using default tag: latest
latest: Pulling from graphile/postgraphile
Digest: sha256:ae708791606842da2aaa8587179a0505c2be1c09820386de471df5319ba7bcc7
Status: Image is up to date for graphile/postgraphile:latest
docker.io/graphile/postgraphile:latest
Sending build context to Docker daemon  6.144kB
Step 1/2 : FROM graphile/postgraphile:latest
 ---> 8c2f0665a259
Step 2/2 : ADD .postgraphilerc.js /postgraphile/
 ---> Using cache
 ---> f9b87b3c8cc0
Successfully built f9b87b3c8cc0
Successfully tagged speller/gql:latest
d32c71dc2bd55d384c0332bf45ef4d784f2c1ace7f4f7abec655855286a02902
Waiting 10 seconds for Docker to come up




Mon  6 Apr 13:06:32 BST 2020
Slept 0 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 12:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 13:26:32 BST 2020
Slept 20 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 12:26:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 13:46:32 BST 2020
Slept 40 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 12:46:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 14:06:32 BST 2020
Slept 60 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 13:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 14:26:32 BST 2020
Slept 80 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 13:26:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 14:46:32 BST 2020
Slept 100 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 13:46:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 15:06:32 BST 2020
Slept 120 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 14:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 15:26:32 BST 2020
Slept 140 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 14:26:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 15:46:32 BST 2020
Slept 160 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 14:46:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 16:06:32 BST 2020
Slept 180 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 15:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 16:26:32 BST 2020
Slept 200 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 15:26:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 16:46:32 BST 2020
Slept 220 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 15:46:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 17:06:32 BST 2020
Slept 240 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 16:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 17:26:32 BST 2020
Slept 260 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 16:26:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 17:46:32 BST 2020
Slept 280 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 16:46:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Mon  6 Apr 18:06:32 BST 2020
Slept 300 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Apr 2020 17:06:32 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}speller-graphql
speller-graphql
DROP DATABASE
DROP ROLE
DROP ROLE
DROP ROLE

```

Everything ran fine for 300 minutes (5 hours). Cannot reproduce.

**Comment by benjie** [comment]
Closing this as I've not heard any other user having this issue and cannot reproduce it - it doesn't seem to be actionable.

**Comment by speller** [comment]
This still happens with me. I'm using Docker for Windows. Just switched it to WSL2 and the same results:

```
0 error(s) as sys_anonymous_test in 282.78ms :: query IntrospectionQuery { __schema { queryType { name } mutationType { name } subscriptionType { name } types { ...FullType } directives { name description locations args { ...InputValue } } } } fragment FullType on __Type { kind name description fields(includeDeprecated: true) { name description args { ...InputValue } type { ...TypeRef } isDeprecated deprecationReason } inputFields { ...InputValue } interfaces { ...TypeRef } enumValues(includeDeprecated: true) { name description isDeprecated deprecationReason } possibleTypes { ...TypeRef } } fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue } fragment TypeRef on __Type { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name } } } } } } } }
Error occurred for PG watching client; reconnecting in 2 seconds. Connection terminated unexpectedly
  postgraphile:postgres listen postgraphile_watch +6m
Error occurred for PG watching client; reconnecting in 2 seconds. Connection terminated unexpectedly
```

```
$ ./run.sh
speller-graphql
speller-graphql
DROP DATABASE
DROP ROLE
DROP ROLE
DROP ROLE
CREATE ROLE
CREATE ROLE
CREATE ROLE
GRANT ROLE
CREATE DATABASE
CREATE SCHEMA
GRANT
CREATE TABLE
GRANT
INSERT 0 3
Using default tag: latest
latest: Pulling from graphile/postgraphile
Digest: sha256:ae708791606842da2aaa8587179a0505c2be1c09820386de471df5319ba7bcc7
Status: Image is up to date for graphile/postgraphile:latest
docker.io/graphile/postgraphile:latest
Sending build context to Docker daemon  6.656kB
Step 1/2 : FROM graphile/postgraphile:latest
 ---> 8c2f0665a259
Step 2/2 : ADD .postgraphilerc.js /postgraphile/
 ---> Using cache
 ---> fc1bc0d911c8
Successfully built fc1bc0d911c8
Successfully tagged speller/gql-test:latest
5c281569083a4b98671f2883b42b7b11c579e85c9c0822ec616e3a367d7e14d2
Waiting 10 seconds for Docker to come up




Tue Apr  7 22:03:19 JST 2020
Slept 0 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Tue, 07 Apr 2020 13:03:19 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Tue Apr  7 22:23:19 JST 2020
Slept 20 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Tue, 07 Apr 2020 13:23:20 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Tue Apr  7 22:43:20 JST 2020
Slept 40 minutes, running request


HTTP/1.1 200 OK
X-GraphQL-Event-Stream: /graphql/stream
Content-Type: application/json; charset=utf-8
Date: Tue, 07 Apr 2020 13:43:20 GMT
Connection: keep-alive
Content-Length: 40

{"data":{"fooById":{"id":1,"text":"A"}}}



Tue Apr  7 23:03:20 JST 2020
Slept 60 minutes, running request


curl: (52) Empty reply from server
```

This time my PC went sleeping. Had to an opportunity to test it without sleeping with your script.

**Comment by benjie** [comment]
Cool; when you retry without sleeping if you can reproduce it you can re-open but I'm really not sure what I can do to fix the issue since I cannot reproduce it.

**Comment by eigood** [comment]
This is happening for us as well, we left a postgraphile with subscriptions-lds enabled, and now get this in our k8s pod.

**Comment by eigood** [comment]
So, in my deployment of postgraphile, I did a grep in all of node_modules for the string 'watching client'.  There is a *single* match.  You can find it at the original location here: https://github.com/graphile/graphile-engine/blob/a58b44211424a194378a677a91e895c33507cdc6/packages/graphile-build-pg/src/plugins/PgIntrospectionPlugin.js#L1206

**Comment by eigood** [comment]
```
2020-12-31 21:20:14.006 UTC [111] keycloak@keycloak LOG:  could not receive data from client: Connection reset by peer
2020-12-31 21:20:14.006 UTC [11389] moqui@moqui LOG:  could not receive data from client: Connection reset by peer
2020-12-31 21:20:15.210 UTC [11429] postgres@moqui ERROR:  relation "postgraphile_meta.logical_decoding_slots" does not exist at character 60
2020-12-31 21:20:15.210 UTC [11429] postgres@moqui STATEMENT:  
                  with deleted_slots as (
                    delete from postgraphile_meta.logical_decoding_slots
                    where last_checkin < now() - interval '1 hour'
                    returning *
                  )
                  select pg_catalog.pg_drop_replication_slot(slot_name)
                  from deleted_slots
                  where exists (
                    select 1
                    from pg_catalog.pg_replication_slots
                    where pg_replication_slots.slot_name = deleted_slots.slot_name
                  )
```

**Comment by speller** [comment]
I believe the issue is caused by the WSL date bug: https://github.com/microsoft/WSL/issues/5324 . I faced the hanging issue when I was running Postgres outside of the Docker and Postgraphile inside. When I moved both into the Docker, I never saw hanging.

**Comment by benjie** [comment]
Whatever it is, I feel like it's probably not a bug in PostGraphile itself, so I'm leaving this closed. If someone comes up with reliable reproduction steps that show it is an issue in PostGraphile, please let me know :+1:

**Comment by benjie** [comment]
This turned out to be a bug in watch mode which you only experience when you have a lot of unexpectedly terminated PostgreSQL connections (9 for default PostGraphile options). This is why sleeping would have triggered it, I expect - because the connections would time out. This is fixed in `4.11` which I just released; PR:  https://github.com/graphile/graphile-engine/pull/711

Thanks to @eigood for tracking it down!
